triples-generator/src/main/scala/io/renku/triplesgenerator/events/categories/cleanup/ProjectTriplesRemover.scala

```diff
@@ -0,0 +1,177 @@
+import io.renku.rdfstore.SparqlQuery.Prefixes
+import io.renku.rdfstore.{RdfStoreClientImpl, RdfStoreConfig, SparqlQuery, SparqlQueryTimeRecorder}
+import org.typelevel.log4cats.Logger
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+private trait ProjectTriplesRemover[F[_]] {
+  def removeTriples(of: projects.Path): F[Unit]
+}
+
+private object ProjectTriplesRemover {
+  def apply[F[_]: Async: Logger](
+      timeRecorder:   SparqlQueryTimeRecorder[F],
+      retryInterval:  FiniteDuration = SleepAfterConnectionIssue,
+      maxRetries:     Int Refined NonNegative = MaxRetriesAfterConnectionTimeout,
+      idleTimeout:    Duration = 6 minutes,
+      requestTimeout: Duration = 5 minutes
+  ): F[ProjectTriplesRemover[F]] = for {
+    rdfStoreConfig <- RdfStoreConfig[F]()
+    renkuBaseUrl   <- RenkuBaseUrlLoader[F]()
+  } yield new ProjectTriplesRemoverImpl[F](rdfStoreConfig,
+                                           timeRecorder,
+                                           renkuBaseUrl,
+                                           retryInterval,
+                                           maxRetries,
+                                           idleTimeout,
+                                           requestTimeout
+  )
+}
+private class ProjectTriplesRemoverImpl[F[_]: Async: Logger](
+    rdfStoreConfig: RdfStoreConfig,
+    timeRecorder:   SparqlQueryTimeRecorder[F],
+    renkuBaseUrl:   RenkuBaseUrl,
+    retryInterval:  FiniteDuration = SleepAfterConnectionIssue,
+    maxRetries:     Int Refined NonNegative = MaxRetriesAfterConnectionTimeout,
+    idleTimeout:    Duration = 6 minutes,
+    requestTimeout: Duration = 5 minutes
+) extends RdfStoreClientImpl(rdfStoreConfig,
+                             timeRecorder,
+                             retryInterval = retryInterval,
+                             maxRetries = maxRetries,
+                             idleTimeoutOverride = idleTimeout.some,
+                             requestTimeoutOverride = requestTimeout.some
+    )
+    with ProjectTriplesRemover[F] {
+  implicit val baseUrl = renkuBaseUrl
+
+  override def removeTriples(projectPath: projects.Path): F[Unit] = for {
+    _ <- relinkDatasetHierarchy(projectPath)
+    _ <- relinkProjectHierarchy(projectPath)
+    _ <- removeProjectInternalEntities(projectPath)
+  } yield ()
+
+  private def relinkDatasetHierarchy(projectPath: projects.Path): F[Unit] = for {
+    _ <- updateWithNoResult(datasetSameAsUpdate(projectPath))
+    _ <- updateWithNoResult(datasetTopmostSameAsUpdate(projectPath))
+  } yield ()
+
+  private def datasetSameAsUpdate(projectPath: projects.Path): SparqlQuery = SparqlQuery.of(
+    name = "CLEAN UP - dataset re-linking topmostSameAs hierarchy",
+    prefixes,
+    s"""
+      INSERT {
+        ?directDescendant schema:sameAs ?sameAs .
+      }
+      WHERE {
+        ?dataset ^renku:hasDataset <$renkuBaseUrl/projects/$projectPath>;
```

Comment:
Would it be possible to use the `renku:projectPath` on the Project entity instead of constructing the resourceId? Then, we could be even  able to get rid of the `renkuBaseUrl` in this class.

Reply:
Not sure if this is really worth it as we will add another union(merge join) to the query. As the queries in this PR can be costly I wouldn't sacrifice performance here.